### ISSUE TITLE
Details of perceptual path regularization

### ISSUE THREAD

**Issue by kylepob** [rosinality/stylegan2-pytorch#48]
Hello, thank you for this amazing repo! I had a couple of questions related to the perceptual path regularization. 

1) Could you give an explanation of the arguments "path_regularize" and "path_batch_shrink" and why they are both set to a default of 2?

2) Would you mind explaining the purpose of the **bolded** summing operation on line 227 in train.py?

weighted_path_loss = args.path_regularize * args.g_reg_every * path_loss
if args.path_batch_shrink:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**weighted_path_loss += 0 * fake_img[0, 0, 0, 0]**

**Comment by rosinality** [comment]
1. path_regularize is the weight of path length regularization term. path_batch_shrink is reducing factor (so, in default, half) of batch sizes when calculating path regularization term. Mainly for reducing memory consumptions, I think.
2. It is to trigger gradient allreduce of DistributedDataParallel. If you only use losses calculated using autograd.grad, gradient allreduce will not be triggered so gradients of each worker processes will be different. So I added auxiliary terms that will not affect gradients to trigger allreduce.

**Comment by kylepob** [comment]
Thank you!

**Comment by TropComplique** [comment]
Hi @rosinality.
I really like your implementation of StyleGAN2.

I think that `weighted_path_loss += 0 * fake_img[0, 0, 0, 0]` is not needed in pytorch 1.6 
because gradient allreduce is triggered and gradients on each worker are the same.

Also `+ 0 * real_pred[0]` is not needed [here](https://github.com/rosinality/stylegan2-pytorch/blob/master/train.py#L223).

Here is a small example that shows gradients on each worker:  
```python
import os
import torch
import torch.nn as nn

import torch.distributed as dist
import torch.multiprocessing as mp


NUM_GPUS = 3


class Model(nn.Module):

    def __init__(self):
        super(Model, self).__init__()
        self.w = nn.Parameter(torch.rand(1))

    def forward(self, x):
        return self.w * x.pow(2)


def worker(rank):

    torch.cuda.set_device(rank)
    device = torch.device(rank)
    dist.init_process_group(backend='nccl', world_size=NUM_GPUS, rank=rank)

    def parallelize(module):
        return nn.parallel.DistributedDataParallel(module, device_ids=[rank])

    model = Model().to(device)
    model = parallelize(model)

    # all workers have the same initial model
    w = model.module.w
    print(f'initial weights at {rank}:', w.data)

    x = torch.randn(3).to(device)
    x.requires_grad = True
    y = model(x)  # shape [3]

    # all workers have different data
    print(f'input data at {rank}:', x)

    grad = torch.autograd.grad(y.sum(), x, create_graph=True)[0]
    loss = grad.pow(2).mean(0)  # gradient penalty

    # compare with gradient calculated by hand
    assert torch.isclose(2 * x * w, grad).all()

    model.zero_grad()
    loss.backward()

    # all workers have the same grad
    print(f'final gradient at {rank}:', w.grad)

    # compare with gradient calculated by hand
    t = (8 * x.pow(2) * w).mean(0)
    dist.all_reduce(t, op=dist.ReduceOp.SUM)
    assert torch.isclose(t/NUM_GPUS, w.grad).all()


def main():
    os.environ['MASTER_ADDR'] = '127.0.0.1'
    os.environ['MASTER_PORT'] = '29500'
    mp.spawn(worker, nprocs=NUM_GPUS, args=())


if __name__ == '__main__':
    main()
```

What do you think?

**Comment by rosinality** [comment]
@TropComplique Oh, it's nice! But it could be better to make it conditional using version check for (some) compatibilities.

**Comment by FeryET** [comment]
Sorry I have a question regarding dimensionality of PL grads. In this section:

https://github.com/rosinality/stylegan2-pytorch/blob/bef283a1c24087da704d16c30abc8e36e63efa0e/train.py#L87-L100

You compute the grads which should be a 2D tensor, and then sum it on the dim=2 and dim=1. I'm trying to implement stylegan2 and by reading your code I'm confused. Can you elaborate what is happening here? If you could start from the dimensionality of inputs and explain the dimensionality of each line I would really really really appreciate it.

**Comment by rosinality** [comment]
@FeryET Actually latents is 3d tensor which has dimensions [batch, n_latents, dim]. Latents for each layers could be different as we uses style mixing regularizations, so instead of 1 latents multiple latents used in here. I think dimensions of other tensors will be now straightforward.

**Comment by FeryET** [comment]
> @FeryET Actually latents is 3d tensor which has dimensions [batch, n_latents, dim]. Latents for each layers could be different as we uses style mixing regularizations, so instead of 1 latents multiple latents used in here. I think dimensions of other tensors will be now straightforward.

Somehow I forgot about style mixing! Thanks a lot.
